src/index.js

```diff
@@ -0,0 +1,269 @@
+    this._dials = new Set()
+
+    this._onConnection = this._onConnection.bind(this)
+    this._dialPeer = this._dialPeer.bind(this)
+  }
+
+  _addPeer (peer) {
+    const id = peer.info.id.toB58String()
+
+    /*
+      Always use an existing peer.
+
+      What is happening here is: "If the other peer has already dialed to me, we already have
+      an establish link between the two, what might be missing is a
+      Connection specifically between me and that Peer"
+     */
+    let existing = this.peers.get(id)
+    if (!existing) {
+      this.log('new peer', id)
+      this.peers.set(id, peer)
+      existing = peer
+
+      peer.once('close', () => this._removePeer(peer))
+    }
+    ++existing._references
+
+    return existing
+  }
+
+  _removePeer (peer) {
+    const id = peer.info.id.toB58String()
+
+    this.log('remove', id, peer._references)
+    // Only delete when no one else is referencing this peer.
+    if (--peer._references === 0) {
+      this.log('delete peer', id)
+      this.peers.delete(id)
+    }
+
+    return peer
+  }
+
+  _dialPeer (peerInfo, callback) {
+    callback = callback || function noop () { }
+    const idB58Str = peerInfo.id.toB58String()
+
+    // If already have a PubSub conn, ignore
+    const peer = this.peers.get(idB58Str)
+    if (peer && peer.isConnected) {
+      return nextTick(() => callback())
+    }
+
+    // If already dialing this peer, ignore
+    if (this._dials.has(idB58Str)) {
+      this.log('already dialing %s, ignoring dial attempt', idB58Str)
+      return nextTick(() => callback())
+    }
+    this._dials.add(idB58Str)
+
+    this.log('dialing %s', idB58Str)
+    this.libp2p.dialProtocol(peerInfo, this.multicodec, (err, conn) => {
+      this.log('dial to %s complete', idB58Str)
+
+      // If the dial is not in the set, it means that pubsub has been
+      // stopped
+      const pubsubStopped = !this._dials.has(idB58Str)
+      this._dials.delete(idB58Str)
+
+      if (err) {
+        this.log.err(err)
+        return callback()
+      }
+
+      // pubsub has been stopped, so we should just bail out
+      if (pubsubStopped) {
+        this.log('pubsub was stopped, not processing dial to %s', idB58Str)
+        return callback()
+      }
+
+      this._onDial(peerInfo, conn, callback)
+    })
+  }
+
+  _onDial (peerInfo, conn, callback) {
+    const idB58Str = peerInfo.id.toB58String()
+    this.log('connected', idB58Str)
+
+    const peer = this._addPeer(new Peer(peerInfo))
+    peer.attachConnection(conn)
+
+    nextTick(() => callback())
+  }
+
+  _onConnection (protocol, conn) {
+    conn.getPeerInfo((err, peerInfo) => {
+      if (err) {
+        this.log.err('Failed to identify incomming conn', err)
+        return pull(empty(), conn)
+      }
+
+      const idB58Str = peerInfo.id.toB58String()
+      const peer = this._addPeer(new Peer(peerInfo))
+
+      this._processConnection(idB58Str, conn, peer)
+    })
+  }
+
+  /**
+   * Process connections
```

Comment:
`Process connections` isn't helpful to implementors. What kind of actions should be taken here when overriding? When is this called and by whom?  The function is pretty vague so I think we should provide some additional details of what kind of things implementors might do here.